http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/AppUsageComponent.scala

```diff
@@ -0,0 +1,107 @@
+package org.broadinstitute.dsde.workbench.leonardo
+package db
+
+import cats.effect.Sync
+import cats.implicits._
+import org.broadinstitute.dsde.workbench.leonardo.AppId
+import org.broadinstitute.dsde.workbench.leonardo.db.LeoProfile.api._
+import org.broadinstitute.dsde.workbench.leonardo.db.LeoProfile.dummyDate
+import org.broadinstitute.dsde.workbench.leonardo.db.LeoProfile.mappedColumnImplicits._
+import org.broadinstitute.dsde.workbench.leonardo.http.dbioToIO
+import org.broadinstitute.dsde.workbench.openTelemetry.OpenTelemetryMetrics
+import org.typelevel.log4cats.Logger
+
+import java.sql.SQLDataException
+import java.time.Instant
+import scala.concurrent.ExecutionContext
+
+class AppUasageTable(tag: Tag) extends Table[AppUsageRecord](tag, "APP_USAGE") {
+  def id = column[AppUsageId]("id", O.PrimaryKey, O.AutoInc)
+  def appId = column[AppId]("appId")
+
+  def startTime = column[Instant]("startTime", O.SqlType("TIMESTAMP(6)"))
+  def stopTime = column[Instant]("stopTime", O.SqlType("TIMESTAMP(6)"))
+
+  def * = (id, appId, startTime, stopTime) <> (AppUsageRecord.tupled, AppUsageRecord.unapply)
+}
+
+object appUsageQuery extends TableQuery(new AppUasageTable(_)) {
+  def recordStart[F[_]](appId: AppId, startTime: Instant)(implicit
+    ec: ExecutionContext,
+    dbReference: DbReference[F],
+    metrics: OpenTelemetryMetrics[F],
+    F: Sync[F],
+    logger: Logger[F]
+  ): F[AppUsageId] = {
+
+    val dbio = for {
+      recordExists <- appUsageQuery
+        .filter(_.appId === appId)
+        .filter(_.stopTime === dummyDate)
+        .exists
+        .result // Make sure there's no existing appId that doesn't have endTime recorded already
+      id <-
+        if (recordExists)
+          DBIO.failed(
+            new SQLDataException(s"app(${appId.id}) usage startTime was recorded previously with no endTime recorded")
+          )
+        else
+          appUsageQuery returning appUsageQuery.map(_.id) += AppUsageRecord(AppUsageId(-1), appId, startTime, dummyDate)
+    } yield id
+
+    for {
+      res <- dbio.transaction.attempt
+      id <- res match {
+        case Left(e) if e.getMessage.contains("usage startTime was recorded previously") =>
+          // We should alert if this happens
```

Comment:
thinking of creating alerts once this PR is merged, and revisiting proper way of addressing the case then

Reply:
Could you elaborate a bit on this failure case? I am not sure I understand it based on the error message above

Reply:
this is when we have a record already for this app and that record doesn't have `stopTime` recorded.

so imagine in DB we have the following already, we should error if leo tries to record another startTime

| id | appId | startTime                  | endTime                    |
|----|-------|----------------------------|----------------------------|
| 1  | 100   | 2018-01-24 20:48:39.666000 | 1970-01-01 00:00:01.000000 |

Reply:
Ohhhh got it, makes complete sense, thanks!